File: agentarmy-dashboard/orchestration/education_center/education_center.py
"""
EducationCenter top-level domain implementation.
"""
from copy import deepcopy
from typing import Any, Dict

from .assessment_agent import AssessmentAgent
from .curriculum_agent import CurriculumAgent
from .events import (
    EDU_ASSESSMENT_COMPLETED,
    EDU_LESSON_GENERATED,
    EDU_PROGRESS_UPDATED,
    EDU_REWARD_ISSUED,
    EDU_SESSION_BLOCKED,
    EDU_SESSION_STARTED,
    EDU_SIMULATION_REQUESTED,
)
from .knowledge_agent import KnowledgeAgent
from .learning_style_agent import LearningStyleAgent
from .policies import age_band, classify_topic, get_default_policies
from .progress_agent import ProgressAgent
from .runtime_hooks import EducationRuntimeHooks
from .safety_agent import SafetyAgent
from .simulation_agent import SimulationAgent


def _safe_int(value, default=0):
    try:
        return int(value)
    except Exception:
        return default


class EducationCenter:
    def attach_event_bus(self, event_bus):
        """No-op for compatibility with agent registration."""
        pass
    """
    Sovereign learning domain coordinating six teaching agents + safety enforcement.
    """

    def __init__(self, runtime):
        self.runtime = runtime
        self.hooks = EducationRuntimeHooks(runtime)
        self.policies = get_default_policies()
        self.state = {
            "learners": {},
            "sessions": {},
            "events": [],
        }

        self.knowledge_agent = KnowledgeAgent(self)
        self.curriculum_agent = CurriculumAgent(self)
        self.assessment_agent = AssessmentAgent(self)
        self.simulation_agent = SimulationAgent(self)
        self.learning_style_agent = LearningStyleAgent(self)
        self.progress_agent = ProgressAgent(self)
        self.safety_agent = SafetyAgent(self)
        self.agents = [
            self.knowledge_agent,
            self.curriculum_agent,
            self.assessment_agent,
            self.simulation_agent,
            self.learning_style_agent,
            self.progress_agent,
        ]

    def bootstrap(self):
        self.register_with_runtime()
        return {
            "domain": "education_center",
            "registered_agents": [agent.name for agent in self.agents],
            "policies_version": self.policies.get("version"),
        }

    def register_with_runtime(self):
        if hasattr(self.runtime, "register_domain"):
            self.runtime.register_domain("education_center", self)
        if hasattr(self.runtime, "register_agent"):
            for agent in self.agents:
                self.runtime.register_agent(agent.name, agent)

    def _publish(self, event_type: str, payload: Dict[str, Any]):
        event = {"type": event_type, "payload": payload}
        self.state["events"] = [event, *self.state["events"]][:500]
        event_bus = getattr(self.runtime, "event_bus", None)
        if event_bus and hasattr(event_bus, "publish"):
            from orchestration.runtime_core.event_bus import Event

            event_bus.publish(Event(event_type, payload))

    def _learner(self, learner_id: str, profile: Dict[str, Any]):
        if current := self.state["learners"].get(learner_id):
            return current
        age = _safe_int(profile.get("age", 18), 18)
        learner = {
            "learner_id": learner_id,
            "age": age,
            "age_band": age_band(age, self.policies),
            "level": str(profile.get("level", "beginner")),
            "style": str(profile.get("style", "mixed")),
            "mastery": _safe_int(profile.get("mastery", 0), 0),
            "completed_sessions": 0,
            "reward_balance": 0,
        }
        self.state["learners"][learner_id] = learner
        return learner

    def start_session(self, request: Dict[str, Any]):
        learner_id = str(request.get("learner_id") or "anonymous")
        topic = str(request.get("topic") or "general literacy")
        learner = self._learner(learner_id, request.get("learner_profile", {}))
        safety = self.safety_agent.filter_content(topic, learner)
        defensive = self.hooks.defensive_check(safety["topic_safety"], learner)
        governance = self.hooks.governance_check(safety["topic_safety"], learner)

        if not defensive["allowed"] or not governance["allowed"] or not safety["allowed"]:
            payload = {
                "learner_id": learner_id,
                "topic": topic,
                "reason": "blocked_by_policy",
                "safety": safety,
                "defensive": defensive,
                "governance": governance,
            }
            self._publish(EDU_SESSION_BLOCKED, payload)
            return {"status": "blocked", **payload}

        path = self.curriculum_agent.build_path(topic, learner)
        explanation = self.knowledge_agent.explain(topic, learner["level"], learner["style"])
        swarm_plan = self.hooks.swarm_teaching_plan(topic, learner["level"])
        session_id = f"edu-{len(self.state['sessions']) + 1}"
        session = {
            "session_id": session_id,
            "learner_id": learner_id,
            "topic": topic,
            "path": path,
            "swarm_plan": swarm_plan,
            "status": "active",
        }
        self.state["sessions"][session_id] = session

        self._publish(EDU_SESSION_STARTED, {"session_id": session_id, "learner_id": learner_id, "topic": topic})
        self._publish(EDU_LESSON_GENERATED, {"session_id": session_id, "lesson": explanation})
        return {
            "status": "active",
            "session": session,
            "lesson": explanation,
            "policies": {
                "defensive": defensive,
                "governance": governance,
            },
        }

    def submit_assessment(self, request: Dict[str, Any]):
        session_id = str(request.get("session_id") or "")
        session = self.state["sessions"].get(session_id)
        if not session:
            return {"status": "failed", "error": "session_not_found"}
        learner = self.state["learners"][session["learner_id"]]
        score = self.assessment_agent.score_submission(request.get("submission", {}))
        reward = self.hooks.economic_reward(score, self.policies["reward_policy"])

        learner["mastery"] = max(0, min(100, learner["mastery"] + int(score / 10)))
        if score >= self.policies["reward_policy"]["pass_threshold"]:
            learner["completed_sessions"] += 1
            learner["reward_balance"] += reward
            session["status"] = "completed"

        progress = self.progress_agent.track(learner)
        adaptation = self.learning_style_agent.adapt(learner, {"confused": score < 50})

        payload = {
            "session_id": session_id,
            "learner_id": learner["learner_id"],
            "score": score,
            "progress": progress,
            "adaptation": adaptation,
        }
        self._publish(EDU_ASSESSMENT_COMPLETED, payload)
        self._publish(EDU_PROGRESS_UPDATED, {"learner_id": learner["learner_id"], "progress": progress})
        if reward > 0:
            self._publish(EDU_REWARD_ISSUED, {"learner_id": learner["learner_id"], "reward": reward})
        return {"status": "recorded", "score": score, "reward": reward, "progress": progress, "adaptation": adaptation}

    def generate_simulation(self, request: Dict[str, Any]):
        session_id = str(request.get("session_id") or "")
        session = self.state["sessions"].get(session_id)
        if not session:
            return {"status": "failed", "error": "session_not_found"}
        learner = self.state["learners"][session["learner_id"]]
        simulation = self.simulation_agent.simulate(session["topic"], learner)
        self._publish(EDU_SIMULATION_REQUESTED, {"session_id": session_id, "simulation": simulation})
        return {"status": "ok", "simulation": simulation}

    def get_policy(self):
        return deepcopy(self.policies)

    def get_state(self):
        return {
            "domain": "education_center",
            "learners": list(self.state["learners"].values()),
            "sessions": list(self.state["sessions"].values()),
            "events": list(self.state["events"][:50]),
            "policy_version": self.policies.get("version"),
        }

File: agentarmy-dashboard/server/educationCenter.js
const crypto = require('node:crypto');

const EDUCATION_STATE_KEY = 'education_center_v1';

function nowIso() {
  return new Date().toISOString();
}

function normalizeText(value, fallback = '') {
  const text = String(value || '').trim();
  return text || fallback;
}

function normalizeNumber(value, fallback = 0) {
  const num = Number(value);
  return Number.isFinite(num) ? num : fallback;
}

function ageBand(age) {
  if (age <= 0) return 'unknown';
  if (age < 13) return 'child';
  if (age < 18) return 'teen';
  return 'adult';
}

function topicSafety(topic = '') {
  const lower = normalizeText(topic).toLowerCase();
  const restricted = ['exploit', 'malware', 'weapon', 'self-harm', 'fraud', 'illegal'];
  const mature = ['violence', 'sexual', 'substance', 'gambling'];
  if (restricted.some((k) => lower.includes(k))) return { level: 'blocked', reason: 'restricted topic' };
  if (mature.some((k) => lower.includes(k))) return { level: 'mature', reason: 'mature topic' };
  return { level: 'safe', reason: 'no restrictions' };
}

function loadState(db) {
  return db.config.get(EDUCATION_STATE_KEY) || {
    version: '1.0.0',
    learners: {},
    sessions: [],
    reward_ledger: [],
    updated_at: nowIso(),
  };
}

function saveState(db, state) {
  db.config.set(EDUCATION_STATE_KEY, {
    ...state,
    updated_at: nowIso(),
  });
}

function getOrCreateLearner(state, profile, actor = 'unknown') {
  const learnerId = normalizeText(profile?.learner_id, actor);
  const existing = state.learners[learnerId];
  if (existing) return existing;
  const created = {
    learner_id: learnerId,
    age: normalizeNumber(profile?.age, 18),
    age_band: ageBand(normalizeNumber(profile?.age, 18)),
    mode: normalizeText(profile?.mode, 'general'),
    level: normalizeText(profile?.level, 'beginner'),
    style: normalizeText(profile?.style, 'mixed'),
    mastery_by_topic: {},
    completed_sessions: 0,
    qubitcoin_balance: 0,
    created_at: nowIso(),
    updated_at: nowIso(),
  };
  state.learners[learnerId] = created;
  return created;
}

function buildLearningPath(topic, level = 'beginner') {
  const base = normalizeText(topic, 'general literacy');
  const lvl = normalizeText(level, 'beginner').toLowerCase();
  const first = `Foundations of ${base}`;
  const second = lvl === 'advanced' ? `Applied systems in ${base}` : `Guided practice in ${base}`;
  const third = lvl === 'advanced' ? `Research-grade synthesis for ${base}` : `Capstone for ${base}`;
  return [
    { step_id: 's1', title: first, agent: 'CurriculumAgent', difficulty: 'baseline' },
    { step_id: 's2', title: second, agent: 'KnowledgeAgent', difficulty: lvl === 'advanced' ? 'high' : 'medium' },
    { step_id: 's3', title: third, agent: 'AssessmentAgent', difficulty: lvl === 'advanced' ? 'high' : 'medium' },
  ];
}

function buildLessonCard(session, learner) {
  const modality = learner.style === 'visual' ? 'diagram + examples' : learner.style === 'audio' ? 'spoken walkthrough' : 'mixed text + practice';
  return {
    lesson_id: `lesson-${crypto.randomUUID()}`,
    topic: session.topic,
    explanation: `KnowledgeAgent explains ${session.topic} for ${learner.level} learners using ${modality}.`,
    simulation_prompt: `SimulationAgent: build a sandbox scenario for ${session.topic} with progressive hints.`,
    assessment_prompt: `AssessmentAgent: produce 5 checks focused on ${session.topic} mastery gaps.`,
    generated_at: nowIso(),
  };
}

function startSession(db, payload = {}, actor = 'unknown', options = {}) {
  const state = loadState(db);
  const role = normalizeText(options?.role, 'user');
  const requestedLearnerId = normalizeText(payload?.learner?.learner_id);
  if (role !== 'admin' && requestedLearnerId && requestedLearnerId !== actor) {
    return {
      status: 'blocked',
      message: 'learner scope violation',
    };
  }
  const learnerProfile = {
    ...(payload.learner || {}),
    learner_id: role === 'admin' ? (requestedLearnerId || actor) : actor,
  };
  const learner = getOrCreateLearner(state, learnerProfile, actor);
  const topic = normalizeText(payload.topic, 'general problem solving');
  const safety = topicSafety(topic);

  if (safety.level === 'blocked') {
    return {
      status: 'blocked',
      safety,
      learner,
      message: 'Topic blocked by EducationCenter safety policy.',
    };
  }

  if ((learner.age_band === 'child' || learner.age_band === 'teen') && safety.level === 'mature') {
    return {
      status: 'pending_approval',
      safety,
      learner,
      message: 'Mature topic requires parent/teacher approval.',
    };
  }

  const path = buildLearningPath(topic, learner.level);
  const session = {
    session_id: `edu-${crypto.randomUUID()}`,
    learner_id: learner.learner_id,
    topic,
    goal: normalizeText(payload.goal, `Build usable skill in ${topic}`),
    mode: learner.mode,
    status: 'active',
    path,
    progress_index: 0,
    mastery_estimate: normalizeNumber(learner.mastery_by_topic[topic], 0),
    created_at: nowIso(),
    updated_at: nowIso(),
  };
  const lesson = buildLessonCard(session, learner);
  state.sessions = [session, ...state.sessions].slice(0, 500);
  learner.updated_at = nowIso();
  saveState(db, state);

  return {
    status: 'active',
    safety,
    session,
    lesson,
    hooks: {
      defensive: ['age_filter', 'content_sensitivity', 'safe_mode'],
      governance: ['parent_teacher_approval', 'policy_override_controls'],
      economic: ['qubitcoin_rewards', 'streak_bonus'],
      swarm: ['multi_agent_teaching_team'],
    },
    team: ['KnowledgeAgent', 'CurriculumAgent', 'AssessmentAgent', 'SimulationAgent', 'LearningStyleAgent', 'ProgressAgent', 'SafetyAgent'],
  };
}

function submitAssessment(db, payload = {}, actor = 'unknown', options = {}) {
  const state = loadState(db);
  const role = normalizeText(options?.role, 'user');
  const sessionId = normalizeText(payload.session_id);
  const score = Math.max(0, Math.min(100, normalizeNumber(payload.score, 0)));
  const session = state.sessions.find((s) => s.session_id === sessionId);
  if (!session) throw new Error('session not found');
  if (role !== 'admin' && session.learner_id !== actor) {
    throw new Error('forbidden session access');
  }

  const learner = state.learners[session.learner_id] || getOrCreateLearner(state, { learner_id: session.learner_id }, actor);
  const priorMastery = normalizeNumber(learner.mastery_by_topic[session.topic], 0);
  const delta = Math.round((score / 100) * 20);
  const nextMastery = Math.max(0, Math.min(100, priorMastery + delta));
  learner.mastery_by_topic[session.topic] = nextMastery;
  learner.completed_sessions += score >= 60 ? 1 : 0;

  const reward = score >= 60 ? 10 + Math.round(score / 10) : 0;
  learner.qubitcoin_balance += reward;
  learner.updated_at = nowIso();

  session.mastery_estimate = nextMastery;
  session.progress_index = Math.min(session.path.length, session.progress_index + 1);
  session.status = session.progress_index >= session.path.length ? 'completed' : 'active';
  session.updated_at = nowIso();

  state.reward_ledger = [
    {
      id: `rew-${crypto.randomUUID()}`,
      learner_id: learner.learner_id,
      session_id: session.session_id,
      amount: reward,
      reason: reward > 0 ? 'assessment_mastery' : 'no_reward',
      issued_at: nowIso(),
    },
    ...state.reward_ledger,
  ].slice(0, 1000);

  saveState(db, state);

  return {
    status: 'recorded',
    learner_id: learner.learner_id,
    session_id: session.session_id,
    score,
    mastery_before: priorMastery,
    mastery_after: nextMastery,
    reward_issued: reward,
    session_status: session.status,
    next_step: session.path[session.progress_index] || null,
  };
}

function buildSimulation(payload = {}) {
  const topic = normalizeText(payload.topic, 'general problem solving');
  const level = normalizeText(payload.level, 'beginner');
  return {
    simulation_id: `sim-${crypto.randomUUID()}`,
    topic,
    level,
    scenario: `SimulationAgent scenario: learner navigates a live case on ${topic} at ${level} complexity.`,
    objectives: [
      `Apply core concept of ${topic}`,
      'Make one decision under constraints',
      'Reflect and revise strategy',
    ],
    generated_at: nowIso(),
  };
}

function getStateView(db, options = {}) {
  const state = loadState(db);
  const role = normalizeText(options?.viewerRole, 'user');
  const viewerId = normalizeText(options?.viewerId, 'unknown');
  const allLearners = Object.values(state.learners || {});
  const allSessions = Array.isArray(state.sessions) ? state.sessions : [];
  const allRewards = Array.isArray(state.reward_ledger) ? state.reward_ledger : [];
  const learners = role === 'admin'
    ? allLearners
    : allLearners.filter((l) => normalizeText(l.learner_id) === viewerId);
  const sessions = role === 'admin'
    ? allSessions
    : allSessions.filter((s) => normalizeText(s.learner_id) === viewerId);
  const rewards = role === 'admin'
    ? allRewards
    : allRewards.filter((r) => normalizeText(r.learner_id) === viewerId);
  const active = sessions.filter((s) => s.status === 'active').length;
  const completed = sessions.filter((s) => s.status === 'completed').length;
  return {
    generated_at: nowIso(),
    domain: 'EducationCenter',
    mission: 'Universal adaptive learning domain',
    summary: {
      learners_total: learners.length,
      sessions_total: sessions.length,
      sessions_active: active,
      sessions_completed: completed,
      qubitcoin_issued: rewards.reduce((sum, r) => sum + normalizeNumber(r.amount, 0), 0),
    },
    policies: {
      age_policies: ['child', 'teen', 'adult'],
      safety_policy: 'restricted topics blocked, mature topics require approval for minors',
      progress_policy: 'mastery tracked per topic and advanced via assessments',
      reward_policy: 'QubitCoin issued for passing assessments',
    },
    learners: learners.slice(0, 100),
    recent_sessions: sessions.slice(0, 50),
    recent_rewards: rewards.slice(0, 50),
  };
}

module.exports = {
  startSession,
  submitAssessment,
  buildSimulation,
  getStateView,
  topicSafety,
  ageBand,
};
